modules/apps/frontend-taglib/frontend-taglib-clay/src/main/java/com/liferay/frontend/taglib/clay/servlet/taglib/VerticalNavTag.java

```diff
@@ -118,6 +156,63 @@ protected int processStartTag() throws Exception {
 		return EVAL_BODY_INCLUDE;
 	}
 
+	private void _computeDefaultExpandedKeys(
+		List<String> defaultExpandedKeys,
+		List<VerticalNavItem> verticalNavItems) {
+
+		for (VerticalNavItem verticalNavItem : verticalNavItems) {
+			VerticalNavItemList items =
+				(VerticalNavItemList)verticalNavItem.get("items");
+
+			Boolean expanded = (Boolean)verticalNavItem.get("expanded");
+
+			if (expanded == null) {
+				expanded = Boolean.FALSE;
+			}
+
+			if (expanded) {
+				String itemId = (String)verticalNavItem.get("id");
+
+				if (itemId != null) {
+					defaultExpandedKeys.add(itemId);
+				}
+			}
+
+			if (items != null) {
+				_computeDefaultExpandedKeys(defaultExpandedKeys, items);
+			}
+		}
+	}
```

Comment:
Ooh, recursive algorithms, you don't see this every day! :slightly_smiling_face: